pkg/kubelet/status/state/state_checkpoint_test.go

```diff
@@ -0,0 +1,105 @@
+package state
+
+import (
+	"fmt"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	"k8s.io/kubernetes/pkg/kubelet/checkpointmanager"
+)
+
+func Test_stateCheckpoint_storeState(t *testing.T) {
+	cache := NewStateMemory()
+	checkpointManager, _ := checkpointmanager.NewCheckpointManager(".")
+	checkpointName := "pod_state_checkpoint"
+	sc := &stateCheckpoint{
+		cache:             cache,
+		checkpointManager: checkpointManager,
+		checkpointName:    checkpointName,
+	}
+	var err error
+	type args struct {
+		podResourceAllocation PodResourceAllocation
+	}
+
+	tests := []struct {
+		name string
+		args args
+	}{}
+	suffix := []string{"Ki", "Mi", "Gi", "Ti", "Pi", "Ei", "n", "u", "m", "k", "M", "G", "T", "P", "E", ""}
+	factor := []string{"1", "0.1", "0.03", "10", "100", "512", "1000", "1024", "700", "10000"}
+	for j, fact := range factor {
+		for i, suf := range suffix {
+			if (suf == "E" || suf == "Ei") && (fact == "1000" || fact == "10000") {
+				// when fact is 1000 or 10000, suffix "E" or "Ei", the quantity value is overflow
+				// see detail https://github.com/kubernetes/apimachinery/blob/95b78024e3feada7739b40426690b4f287933fd8/pkg/api/resource/quantity.go#L301
+				continue
+			}
+			tests = append(tests, struct {
+				name string
+				args args
+			}{
+				name: fmt.Sprintf("format %d - cpu %s%s", j*len(suffix)+i, fact, suf),
+				args: args{
+					podResourceAllocation: PodResourceAllocation{
+						"pod1": {
+							"container1": {
+								v1.ResourceCPU:    resource.MustParse(fmt.Sprintf("%s%s", fact, suf)),
+								v1.ResourceMemory: resource.MustParse(fmt.Sprintf("%s%s", fact, suf)),
+							},
+						},
+					},
+				},
+			})
+		}
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			// reset environment
+			err = sc.cache.ClearState()
+			require.NoError(t, err, "failed to clear state")
+			// remove file ./pod_state_checkpoint
+			err = checkpointManager.RemoveCheckpoint(checkpointName)
+			require.NoError(t, err, "failed to remove checkpoint")
+			err = sc.cache.SetPodResourceAllocation(tt.args.podResourceAllocation)
+			require.NoError(t, err, "failed to set pod resource allocation")
+			// store state
+			err = sc.storeState()
+			if err != nil {
+				t.Errorf("failed to store state: %v", err)
+			}
+			// deep copy cache
```

Comment:
Please, use deepcopy-gen, e.g. https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/cm/dra/state/state.go#L26

Reply:
Actually, `sc.cache.GetPodResourceAllocation()` and `sc.cache.GetResizeStatus()` already had the deep copy capability.
```golang
func (s *stateMemory) GetPodResourceAllocation() PodResourceAllocation {
	s.RLock()
	defer s.RUnlock()
	return s.podAllocation.Clone()
}
// Clone returns a copy of PodResourceAllocation
func (pr PodResourceAllocation) Clone() PodResourceAllocation {
	prCopy := make(PodResourceAllocation)
	for pod := range pr {
		prCopy[pod] = make(map[string]v1.ResourceList)
		for container, alloc := range pr[pod] {
			prCopy[pod][container] = alloc.DeepCopy()
		}
	}
	return prCopy
}
func (s *stateMemory) GetResizeStatus() PodResizeStatus {
	s.RLock()
	defer s.RUnlock()
	prs := make(map[string]v1.PodResizeStatus)
	for k, v := range s.podResizeStatus {
		prs[k] = v
	}
	return prs
}
// PodResizeStatus shows status of desired resize of a pod's containers.
type PodResizeStatus string
```

Reply:
I'm just wondering why do you need to clone the whole cache if you only want to veryfy that `PodResourceAllocation` data is the same as `tt.args.podResourceAllocation` after restoring state? Wouldn't it be enough to simply compare `restoredCache.GetPodResourceAllocation())` with `tt.args.podResourceAllocation` ?

Reply:
Yeah, you are right. I should only compare them only. Let me update